src/Processors/Formats/RowInputFormatWithNamesAndTypes.cpp

```diff
@@ -38,7 +38,7 @@ void RowInputFormatWithNamesAndTypes::readPrefix()
     if (getCurrentUnitNumber() != 0)
         return;
 
-    if (with_names || with_types || data_types.at(0)->textCanContainOnlyValidUTF8())
+    if (with_names || with_types)
```

Comment:
Yes, RowBinary format should not remove BOM, because it is binary, not text.

But the logic of this change looks incorrect. You should somehow check that the format is binary. Currently, it is checking something else.

CSV and TSV should skip BOMs.

Reply:
I thought this code is only for `RowBinary*`, my bad.

I haven't found anything useful in `format_settings` to distinguish binary formats, so added an explicit `is_binary` argument.